patches/@endo+ses-ava+0.2.27.patch

```diff
@@ -0,0 +1,250 @@
++ * @param {unknown[]} args
++ * @param {string} name
++ * @param {Logger=} logError
+  */
+-const logErrorFirst = (func, args, name, logger = defaultLogger) => {
++const logErrorFirst = (func, args, name, logError = console.error) => {
+   let result;
+   try {
+     result = apply(func, undefined, args);
+   } catch (err) {
+-    logger(`THROWN from ${name}:`, err);
++    logError(`THROWN from ${name}:`, err);
+     throw err;
+   }
+   if (isPromise(result)) {
+     return result.then(
+       v => v,
+       reason => {
+-        logger(`REJECTED from ${name}:`, reason);
++        logError(`REJECTED from ${name}:`, reason);
+         return result;
+       },
+     );
+@@ -47,7 +62,7 @@ const logErrorFirst = (func, args, name, logger = defaultLogger) => {
+   }
+ };
+ 
+-const testerMethodsWhitelist = [
++const overrideList = [
+   'after',
+   'afterEach',
+   'before',
+@@ -56,32 +71,35 @@ const testerMethodsWhitelist = [
+   'failing',
+   'serial',
+   'only',
+-  'skip',
+ ];
+ 
+ /**
+- * @param {TesterFunc} testerFunc
+- * @param {Logger} [logger]
+- * @returns {TesterFunc} Not yet frozen!
++ * @template {Function} T
++ * @param {T} testerFunc
++ * @param {Logger} [logError]
++ * @returns {T} Not yet frozen!
+  */
+-const wrapTester = (testerFunc, logger = defaultLogger) => {
+-  /** @type {TesterFunc} */
+-  const testerWrapper = (title, implFunc, ...otherArgs) => {
+-    /** @type {ImplFunc} */
++const augmentLogging = (testerFunc, logError = console.error) => {
++  const augmented = (title, implFunc, ...otherArgs) => {
+     const testFuncWrapper = t => {
+       harden(t);
+-      return logErrorFirst(implFunc, [t, ...otherArgs], 'ava test', logger);
++      return logErrorFirst(implFunc, [t, ...otherArgs], 'ava test', logError);
+     };
+     if (implFunc && implFunc.title) {
+       testFuncWrapper.title = implFunc.title;
+     }
+     return testerFunc(title, testFuncWrapper, ...otherArgs);
+   };
+-  return testerWrapper;
++  // re-use other properties (e.g. `.always`)
++  // https://github.com/endojs/endo/issues/647#issuecomment-809010961
++  Object.assign(augmented, testerFunc);
++  // @ts-expect-error cast
++  return augmented;
+ };
+ 
++// TODO check whether this is still necessary in Ava 4
+ /**
+- * The ava `test` function takes a callback argument of the form
++ * The Ava 3 `test` function takes a callback argument of the form
+  * `t => {...}`. If the outcome of this function indicates an error, either
+  * by throwing or by eventually rejecting a returned promise, ava does its
+  * own peculiar console-like display of this error and its stacktrace.
+@@ -105,25 +123,17 @@ const wrapTester = (testerFunc, logger = defaultLogger) => {
+  * that eventually rejects, the error is first sent to the `console` before
+  * propagating into `rawTest`.
+  *
+- * @param {TesterInterface} avaTest
+- * @param {Logger} [logger]
+- * @returns {TesterInterface}
++ * @template {Function} T Ava `test`
```

Comment:
This `@template` line has more syntax than I recognize. What does it mean?

Reply:
`@template T` you're familiar with. The `{Function}` means T has to be a subtype of `Function`. That's so TS knows that the calls below are valid. The string after `T` is a description for display.

Reply:
Ah. Although I knew that extra tokens after jsdoc/ts declarations are assumed to be prose descriptions, somehow that did not occur to me here. It seemed like more type notation that I didn't know. Thanks for clarifying. Resolving.